Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp

```diff
@@ -1446,15 +1446,45 @@ FloatRect AXIsolatedObject::relativeFrame() const
 
     // Having an empty relative frame at this point means a frame hasn't been cached yet.
     if (relativeFrame.isEmpty()) {
-        // InitialFrameRect stores the correct size, but not position, of the element before it is painted.
-        // We find the position of the nearest painted ancestor to use as the position until the object's frame
-        // is cached during painting.
-        auto* ancestor = Accessibility::findAncestor<AXIsolatedObject>(*this, false, [] (const auto& object) {
-            return object.hasCachedRelativeFrame();
-        });
-        relativeFrame = rectAttributeValue<FloatRect>(AXProperty::InitialFrameRect);
-        if (ancestor && relativeFrame.location() == FloatPoint())
-            relativeFrame.setLocation(ancestor->relativeFrame().location());
+        std::optional<IntRect> rectFromLabels;
+        if (isControl()) {
+            // For controls, we can try to use the frame of any associated labels.
+            auto labels = labeledByObjects();
+            for (const auto& label : labels) {
+                std::optional frame = downcast<AXIsolatedObject>(label)->cachedRelativeFrame();
+                if (!frame)
+                    continue;
+                if (!rectFromLabels)
+                    rectFromLabels = *frame;
+                else if (rectFromLabels->intersects(*frame))
```

Comment:
Does it matter that this new frame here intersects? Or should we be uniting all labels

Reply:
In my opinion, yes, since in theory somebody could have labels in very different parts of the page, so uniting them in that case would result in a giant rect. `labeledByObjects()` includes `<label>`s which are typically located near the control, but also includes `aria-labelledby` labels, which I worry may be less likely to located nearby (e.g. someone does a 1x1px hidden label at the top of the page or something). But maybe this is overly cautious. What do you think?

Reply:
A concern I had was if the first label fell into that bucket then we'd have a problem too, but maybe no way around that

Reply:
Yep, you're totally right! I think in that case it's OK, since this is a fallback for the case where the AT needs the frame for something that is not painted, which should very rarely happen (and only happens in this case because the clipping causes the control to never be painted, which I need to fix in a later PR).